# Import

Import functions allow the users to import a json file (or multiple files) generated by auxiliary
 scripts. These `JSON` files should contain sets of keys that make a
 correspondence between the accession numbers and the results of this 
 auxiliary script as the value of that accession number key.

**However, this is achieved by running a [NextFlow](https://www.nextflow.io/)
pipeline, with detailed instructions on each specific section.**

## Download and install requirements to run the pipeline

### Requirements

* [Java 8](http://www.oracle.com/technetwork/java/javase/downloads/index.html) or higher.
* [Docker](https://docs.docker.com/install/) or [Singularity](http://singularity.lbl.gov/install-linux).
* [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html#installation) - follow the two installation steps (there is no need to read anything else).

### Conda recipe for nextflow

If you prefer you can use this conda recipe for nextflow: [![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square)](http://bioconda.github.io/recipes/nextflow/README.html)

```
conda install nextflow
```

## Mapping results

### How to run

`nextflow run tiagofilipe12/pATLASflow --mapping --reads "your_folder/*.fastq"`

### Importing in pATLAS

After, the run of the script has finished you should have a `JSON` file
(extension `.json`), which can be provided to pATLAS in the specific menu.

## Mash screen results

### How to run

`nextflow run tiagofilipe12/pATLASflow --mash_screen --reads "your_folder/*.fastq"`

### Importing in pATLAS

After, the run of the script has finished you should have a `JSON` file
(extension `.json`), which can be provided to pATLAS in the specific menu.

## Sequence

### How to run

`nextflow run tiagofilipe12/pATLASflow --assembly --fasta "your_folder/*.fasta"`

### Importing in pATLAS

After, the run of the script has finished you should have a `JSON` file
(extension `.json`), which can be provided to pATLAS in the specific menu.